Adding abbreviate implementation for Linux.

stupergenius opened the pull request.

kylef commented on Tests/PathKitTests/PathKitSpec.swift:
```diff
@@ -139,12 +139,20 @@ describe("PathKit") {
 
-#if !os(Linux)
   $0.it("can be abbreviated") {
+#if os(Linux)
```
Since `HOME` will also be set on macOS. I think we can unify the test here so the same test covers both platforms. Would you be able to update the test?

kylef commented on Sources/PathKit.swift:
```diff
@@ -152,8 +152,15 @@ extension Path {
-    // TODO: actually de-normalize the path
-    return self
+    let env = ProcessInfo.processInfo.environment
+    guard let home = env["HOME"], self.path.contains(home) else { return self }
```
I think it would be a good idea to make sure that `HOME` is actually set. Otherwise when abbreviating paths when home is unset (empty string) it may result in any path being abbreviated with `~`.

kylef replied:
`contains`? What if the path was something like `/mnt/backups/home/kyle/test`, this would incorrectly match when `HOME` is set to `/home/kyle/` and the method would return `~/mnt/backupstest`

kylef commented on Sources/PathKit.swift:
```diff
@@ -152,8 +152,15 @@ extension Path {
-    return self
+    let env = ProcessInfo.processInfo.environment
+    guard let home = env["HOME"], self.path.contains(home) else { return self }
+    let withoutHome = Path(self.path.replacingOccurrences(of: home, with: ""))
```
What about if the path was the following: `/home/kyle/backups/home/kyle/some/file` and `HOME` was set to `/home/kyle`? This would incorrectly result in `~/backupssome/file`

stupergenius commented:
Thanks for the review @kylef! Very good points. I've updated the tests to unify between Linux/macOS and I've added your suggestions. The path should now be searched from the beginning only for the home path, and only the first occurrence replaced.

kylef commented on Tests/PathKitTests/PathKitSpec.swift:
```diff
@@ -140,18 +140,16 @@ describe("PathKit") {
-#endif
+    guard let home = env["HOME"] else { throw failure("$HOME must be defined in the environment") }
+    
+    print(Path("\(home)/backups\(home)/foo/bar").abbreviate())
```
I think this was left over from local testing?

kylef commented on Sources/PathKit.swift:
```diff
@@ -152,8 +152,18 @@ extension Path {
+    guard let homeRange = self.path.range(of: home, options: [.anchored, .caseInsensitive]) else { return self }
+    let withoutHome = Path(self.path.replacingCharacters(in: homeRange, with: ""))
+    
+    if withoutHome.path.isEmpty || withoutHome.path == "/" {
```
We should probably use `Path.separator` since this could be different on some platforms

kylef replied:
Note, `Path.separator` doesn't change at the moment, but that's a separate issue.

I'd probably reverse the `#if os()` too, so that this code will be used on freeBSD, windows etc. Could be `#if os(osx) || os(iOS) .. etc` with all Apple platforms. Or perhaps we should just use the same implementation everywhere so for consistency. What do you think?

stupergenius replied:
Hm, good question. A couple thoughts:

1. Is freeBSD not caught by `if os(Linux)`? If so, then adding freeBSD would likely involve changes other than this line (e.g. the first line in the file).
1. We may end up in the same trap if apple decides to release a different OS.
1. I'm not really familiar enough to know if this would work on Windows or not? I believe cygwin defines a similar Linux `Home` structure but I don't know about Windows.

Given the above, I feel like, unless we have a suite of tests able to run on the supported platforms, we aren't able to make informed decisions about platform support. When we get those things, and as Apple adds more platforms, then I think we can revisit, and indeed the rest of the code requires a revisiting anyway. Finally, that doing it this way doesn't necessarily prevent us from supporting newer platforms in the future, and instead allows us to constrain which platforms have special behavior, with the default assumption that as new platforms are added, they will either be added with full Foundation support, or we'll need to handle them as special cases anyway.

kylef replied:
I think that what you have should work on all unix platforms, I think you could probably remove the OS conditional and have the same code-path on all and this would compile everywhere. (there is the problem with case sensitive filesystems though). The more I think about this it would be simpler of we can remove the conditional and have the same behaviour and code everywhere that it would work, not sure what your opinion would be on doing that.

The exception being Windows as I am not sure how that works and if `~` is a valid path component or `HOME` would be set.

Swiftenv is tested on Windows (https://github.com/kylef/swiftenv/commit/9c463927c4c2dad900f6a5a97052c6afd64c77ec) but there are unfortunately no binaries to be able to use Swift. I think that `~` may work on Cygwin though. For FreeBSD, there is no CI I found unfortunately (and also the binary problem).

kylef replied:
Regarding CI for FreeBSD, PS4, other BSDs. I think we can just make our best attempt and writing portable code and leave it down to the users to report problems if PathKit won't compile for them on their platform. They can run the test cases to test compatibility.

I don't think it's possible or feasible for us to be able to CI on every platform.

stupergenius replied:
Yeah, I see that and I agree. I suppose as long as the tests still pass on our CI systems that we have now, then unification of implementation is probably the best course. Ideally Foundation at some point has feature parity between platforms, but I think this is the best plan for the short/mid term.

Thanks for your thoughtful reviews!

kylef replied:
Yes, I just checked and Swift itself only has macOS and Linux CI. So there would easily be breakages in Swift itself on these platforms.

kylef commented on Sources/PathKit.swift:
```diff
@@ -152,8 +152,18 @@ extension Path {
+    } else if withoutHome.isAbsolute {
+        return Path("~" + withoutHome.path)
+    } else {
+        return Path("~/" + withoutHome.path)
```
Could use `Path("~") + withoutHome` too so you don't have to worry about `Path.separator`

kylef commented on Sources/PathKit.swift:
```diff
@@ -152,8 +152,18 @@ extension Path {
-    return self
+    let env = ProcessInfo.processInfo.environment
+    guard let home = env["HOME"], !home.isEmpty else { return self }
+    guard let homeRange = self.path.range(of: home, options: [.anchored, .caseInsensitive]) else { return self }
```
I think `caseInsensitive` can be a problem for case sensitive file systems (usually default on linux). Although, I'm not sure how we can check.

We could add a method to normalise a paths case (similar to https://docs.python.org/3/library/os.path.html#os.path.normcase) which then we can normalise both `HOME` and the path before we check (case sensitively).

rbukovansky commented:
FYI, @kylef @stupergenius 

The correct `~` on Windows is `%homedrive%%homepath%` (environment variables joined together):

```batch
C:\>echo %homedrive%%homepath%
C:\Users\richard.bukovansky
```

We are bound to have a lot's of fun when Swift for Windows is released... :wink:

rbukovansky commented:
@stupergenius @kylef One more thing... Don't forget that on Linux user `root` has `/root` home directory and other users have `/home/<username>`...

stupergenius commented:
Thanks for the clarifications Richard. I have been testing with a root user
as well actually (through a docker image) and the tests still pass.
On Tue, Dec 13, 2016 at 7:19 AM Richard Bukovansky <notifications@github.com>
wrote:

> @stupergenius <https://github.com/stupergenius> @kylef
> <https://github.com/kylef> One more thing... Don't forget that on Linux
> user root has /root home directory and other users have /home/<username>
> ...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kylef/PathKit/pull/27#issuecomment-266749722>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AADhZZeNu7rK-Q3zNX3iMHrmCpVc_KBkks5rHqlagaJpZM4LJkwT>
> .
>

stupergenius commented:
I think the above addresses the review comments. The only hairy bits are detecting case sensitive filesystems. Would like to hear your thoughts.

rbukovansky commented:
@stupergenius @kylef

Just FYI: resourceValues() is Swift on Linux can't produce app package as on macOS, it just compiles, links and builds executable.

And assumption that Linux filesystems are always case-sensitive is right (at least with extX filesystems). I have probably never seen Linux machine running case-insensitive (not sure, but I don't think it's possible).
